src/common/WebLink.js

```diff
@@ -1,48 +1,56 @@
 /* @flow strict-local */
 
-import React, { PureComponent } from 'react';
+import React from 'react';
 
-import type { Dispatch } from '../types';
-import { connect } from '../react-redux';
+import invariant from 'invariant';
+import { useSelector } from '../react-redux';
 import Label from './Label';
 import openLink from '../utils/openLink';
-import { getCurrentRealm } from '../selectors';
+import { tryGetCurrentRealm } from '../selectors';
 import { BRAND_COLOR, createStyleSheet } from '../styles';
 
 type Props = $ReadOnly<{|
-  dispatch: Dispatch,
   label: string,
   href: string,
-  realm: URL,
+
+  // What to use to absolutize relative URLs, if different from the
+  // current realm or we don't expect there to be a current realm.
+  baseUrl?: URL,
 |}>;
 
+const componentStyles = createStyleSheet({
+  link: {
+    marginTop: 10,
+    fontSize: 15,
+    color: BRAND_COLOR,
+    textAlign: 'right',
+  },
+});
+
 /**
  * A button styled like a web link.
- *
- * @prop label - Text of the button.
- * @prop href - URL address to open on press.
- * @prop realm - Current realm. Used if the `href` property is relative.
  */
-class WebLink extends PureComponent<Props> {
-  handlePress = () => {
-    const { realm, href } = this.props;
-    openLink(new URL(href, realm).toString());
-  };
-
-  styles = createStyleSheet({
-    link: {
-      marginTop: 10,
-      fontSize: 15,
-      color: BRAND_COLOR,
-      textAlign: 'right',
-    },
-  });
-
-  render() {
-    return <Label style={this.styles.link} text={this.props.label} onPress={this.handlePress} />;
+export default function WebLink(props: Props) {
+  const baseUrl = useSelector(tryGetCurrentRealm) ?? props.baseUrl;
+
+  let url;
+  try {
+    url = new URL(props.href, baseUrl).toString();
+  } catch (e) {
+    invariant(
+      baseUrl != null,
+      "Can't decide what URL to use for a `WebLink`. Did you mean to provide a `baseUrl`?",
+    );
+    throw e;
```

Comment:
As long as we have total control over the URL that the `WebLink` opens, I think it makes sense to error inside the `render` method. (Now at least we have the error boundary, from #4585, so it won't just be a white screen.) Does that sound right?

If we start using user-defined or even server-defined strings, we may want to postpone this error checking until the user actually presses the link, and then we can show a toast or something if it fails to open.

Reply:
Certainly as long as the `href` is a literal in the caller's source code, that seems like the right behavior. That'll help make sure we spot any failure early, likely before it makes it to users.